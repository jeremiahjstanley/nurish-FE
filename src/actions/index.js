export const addFormulas = (formulas) => ({
  type: 'ADD_FORMULAS',
  formulas
})

export const formulaFetchDataSuccess = (formulas) => ({
  type: 'FORMULA_FETCH_DATA_SUCCESS',
  formulas
})

export const isLoading = (bool) => ({
  type: 'IS_LOADING',
  isLoading: bool
})

export const hasErrored = (bool) => ({
  type: 'HAS_ERRORED',
  hasErrored: bool
})