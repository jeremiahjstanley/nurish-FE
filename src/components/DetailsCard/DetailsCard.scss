
.single-card {
	background-color: rgba(255, 255, 255, 0.2);
	display: grid;
	grid-auto-flow: column;
	grid-template-columns: 45% 45%;
	grid-column-gap: 3%;
	justify-content: center;
	width: 65%;
  border-radius: 30px;
  color: #FFF;
  margin: 0 auto 6rem;
  min-width: 20rem;
  padding: 1.5rem;
  text-align: left;
}

.single-card {
	& ul {
		list-style: none;
		padding: 0;
	}
}

.single-card {
	& h2 {
		font-size: 1.25rem;
		margin: .5rem 0;
	}
}

.formula-title {
	align-self: center;
	font-size: 1.5rem;
	grid-column: 1;
	margin: .25rem;
	text-transform: lowercase;
}

.formula-image {
	grid-column: 1;
	margin: 1rem auto;
	max-width: 20rem;
	width: 100%;
}

.single-formula-description {
	grid-column: 1;
	margin: 0;
}

.nutritional-management-header {
	align-self: center;
	grid-column: 1;
}

.usage-guidelines {
	display: grid;
	grid-auto-flow: column;
	grid-column: 1;
	grid-template-rows: 25% 25% 25% 25%;
}

.features-header {
	grid-column: 1;
}

.features-table {
	display: grid;
	grid-auto-flow: column;
	grid-column: 1;
	grid-template-rows: 25% 25% 25% 25%;
}

.nutritional-information-header {
	align-self: center;
	grid-column: 2;
	grid-row: 1;
	margin: .25rem;
}

.nutritional-information {
	grid-column: 2;
	grid-row: 2;
}

.nutritional-information {
	display: grid;
	grid-auto-flow: row;
}

.nutrient-headers {
	display: grid;
	grid-template-columns: 40% 27% 33%;
	border-bottom: 1.5px solid #FFF;
	margin-bottom: .2rem;none
}

.nutrient-header-1 {
	grid-column: 2;
	text-align: center;
}

.nutrient-header-2 {
	grid-column: 3;
	text-align: right;
}

.nutrient-row {
	display: grid;
	grid-template-columns: 50% 35% 15%;
	grid-auto-flow: row;
	border-bottom: .5px solid #FFF;
	margin: .25rem 0;
}

.serving-size {
	margin: .5rem 0;
}

.ingredients-text {
	grid-column: 2;
	grid-row: 3 / 8;
	text-transform: lowercase;
}