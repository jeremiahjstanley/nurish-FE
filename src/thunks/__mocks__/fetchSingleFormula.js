export const fetchSingleFormula = jest.fn();