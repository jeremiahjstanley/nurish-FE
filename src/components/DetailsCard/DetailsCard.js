import React, { Component } from 'react';
import { connect } from 'react-redux';
import Load from '../Loader/Loader';
import { withRouter, NavLink } from 'react-router-dom';
import { fetchSingleFormula } from '../../thunks/fetchSingleFormula.js';
import PropTypes from 'prop-types';
import './DetailsCard.css';

class DetailsCard extends Component {
  constructor() {
    super()

    this.state = {

    }
  }

  static getDerivedStateFromProps(props, state) {
    if (props.formula) {
      return {
        formula: props.formula
      };
    }
    return state 
  }
  
  async componentDidMount() {
    let id = new URL(window.location.href);
    let url = `https://nurish-app.herokuapp.com/api/v1${id.pathname}`
    await this.props.fetchSingleFormula(url);
    this.props.animateLogo('collapse-logo')
  }

  componentWillUnmount() {
    this.props.animateLogo('expand-logo')
  }
	
	usageGuidelines = () => {
    return this.state.formula.usage.split(',').map(use => {
      return (
          <li>{use}</li>
        );
    });
  };

  featuresTable = () => {
		return Object.keys(this.state.formula.formula_overview).map(feature => {
      if (typeof this.state.formula.formula_overview[feature] === 'string') {
        let regex = /_/gi;
        let featureName = feature.replace(regex, ' ');
  			return (
  				<li>{featureName}: {this.state.formula.formula_overview[feature]}</li>
  			);
      }
		});
  };

  nutritionalHeaders = () => {
  	let nutrients = Object.keys(this.state.formula.nutritional_content.properties);
  	let headers = Object.keys(this.state.formula.nutritional_content.properties[nutrients[0]])
  	return headers.map((nutrientHeader, index, arr) => {
      let className = `nutrient-header-${index + 1}`
  		return (
  			<span className={className}>{nutrientHeader}</span>
  		);
  	});
  };

  nutritionalTable = () => {
    return Object.keys(this.state.formula.nutritional_content.properties).map(nutrient => {
  		if (typeof this.state.formula.nutritional_content.properties[nutrient] === 'object') {
	  		var content = Object.keys(this.state.formula.nutritional_content.properties[nutrient]).map(unit => {
	  			return (
	  				<span>
	  					{this.state.formula.nutritional_content.properties[nutrient][unit]}
	  				</span>
	  			);
	  		});
  		} else {
  			return <span className='serving-size'>{nutrient}: {this.state.formula.nutritional_content.properties[nutrient]}</span>	
  		};

		return (
			<li className='nutrient-row'>
  			<span>
  				{nutrient}:
  			</span>
  				{content}
			</li>
			);
  	});
  };

  render() {
    if (this.state.formula.usage) {
      let url = `https://raw.githubusercontent.com/shebesabrina/Nurish-BE/images/public/formula%20images/${this.props.formula.image}`
      return (
        <div className='details-card'>
          <div className='link-container'>
            <a onClick={this.props.history.goBack}>back</a>
            <NavLink to='/search'>search</NavLink>
            <NavLink to='/filter'>filter</NavLink>
            <NavLink to='/formulas'>browse</NavLink>
          </div>
          <div className='single-card'>
          	<h1 className='formula-title'>{this.props.formula.title}</h1>
            <div className='formula-image' alt=''style={ { backgroundImage: `url(${url})` } }>
            </div>
          	<p className='single-formula-description'>
          		{this.props.formula.description}
          	</p>
            <p className='single-formula-restrictions'>
              <strong>{this.props.formula.restrictions}</strong>
            </p>
          	<h2 className='nutritional-management-header'>
          		for the nutritional management of:
          	</h2>
          	<ul className='usage-guidelines'>
          		{this.usageGuidelines()}
      			</ul>
      			<h2 className='features-header'>
      				features at a glance:
      			</h2>
      			<ul className='features-table'>
      				{this.featuresTable()}
      			</ul>
      			<h2 className='nutritional-information-header'>
      				nutritional information:
      			</h2>
      			<ul className='nutritional-information'>
      				 <span className='nutrient-headers'>{this.nutritionalHeaders()}</span>
      				{this.nutritionalTable()}
      			</ul>
      			<p className='ingredients-text'><span className='ingredients-title'><strong>ingredients: </strong></span>
      				{this.props.formula.ingredients}
      			</p>
          </div>
        </div>
      ); 
    } else 
      return (
        <Load/>
      )
  }
};

export const mapStateToProps = (state) => ({
  formula: state.singleFormula
});

export const mapDispatchToProps = (dispatch) => ({
  fetchSingleFormula: (url) => dispatch(fetchSingleFormula(url))
});

DetailsCard.propTypes = {
  fetchSingleFormula: PropTypes.func,
  formula: PropTypes.object
};

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(DetailsCard));