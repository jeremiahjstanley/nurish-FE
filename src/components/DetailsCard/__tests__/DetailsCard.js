import React from 'react';
import { shallow } from 'enzyme';
import {DetailsCard, mapStateToProps, mapDispatchToProps} from '../DetailsCard.js';
import { fetchSingleFormula } from '../../../thunks/fetchSingleFormula.js';
import * as action from '../../../actions';

jest.mock('../../../thunks/fetchSingleFormula');

describe('DetailsCard', () => {
  let wrapper
  let mockformula
  let mockFetchSingleFormula
  let mockhistory
  let mockAnimateLogo



  beforeEach(() => {
    mockhistory = []
    mockAnimateLogo = jest.fn()
    mockFetchSingleFormula = jest.fn()
    mockformula = {nutritional_content: {properties:{nutrient: 'sodium'}, formula_overview: 'best ever', usage: 'thirst'}}
    wrapper = shallow(<DetailsCard fetchSingleFormula={mockFetchSingleFormula} formula={mockformula} history={mockhistory} animateLogo={mockAnimateLogo} />)
  });

  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should return a formula object', () => {
      const mockState = {
      formula: undefined
    };

    const expectedProps = {
      formula: undefined
    };

    const mappedProps = mapStateToProps(mockState);
    expect(mappedProps).toEqual(expectedProps);
  });

  it('should call dispatch with fetchSingleFormula', () => {
    const mockDispatch = jest.fn();
    const mappedProps = mapDispatchToProps(mockDispatch);
    const actionToDispatch = fetchSingleFormula('www.fakeURL');
    mappedProps.fetchSingleFormula('www.fakeURL');
    expect(mockDispatch).toBeCalledWith(actionToDispatch);
  })
  
})