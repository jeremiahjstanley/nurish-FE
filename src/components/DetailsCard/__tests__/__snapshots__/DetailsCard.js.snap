// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DetailsCard should match the snapshot 1`] = `undefined`;
